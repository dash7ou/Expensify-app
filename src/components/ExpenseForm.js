import React from "react";
import moment from "moment";
import {SingleDatePicker} from "react-dates";

class ExpenseForm extends React.Component{
    constructor(props){
        super(props);

        this.state ={
            description: props.expense ? props.expense.description : "",
            note: props.expense ? props.expense.note: '',
            amount: props.expense? (props.expense.amount/100).toString() : "",
            createdAt: props.expense? moment(props.expense.createdAt):moment(),
            focused: false,
            error: undefined
        }
    }

    onDescriptionChange = (e)=>{
        const description = e.target.value;
        this.setState(()=>({description}));
    }
    onChangeNote = (e)=>{
        const note = e.target.value;
        this.setState(()=>({note}))
    }

    onChangeAmount = (e)=>{
        const amount = e.target.value;
        if(!amount || amount.match(/^\d+(\.\d{0,2})?$/)){
            this.setState(()=>({amount}))
        }
    }

    onDateChange = (createdAt)=>{
        if(createdAt){
            this.setState(()=>({createdAt}))
        }
    }

    onFocusChange = ({focused})=>{
        this.setState(()=>({focused}))
    }

    onSubmit = (e)=>{
        e.preventDefault();

        if(!this.state.description){
            return this.setState(()=>({error:"Add description please!"}))
        }

        if(!this.state.amount){
            return this.setState(()=>({error: "Add amount Please!"}))
        }

        this.setState(()=>({error: ""}))
        this.props.onSubmit({
            description: this.state.description,
            amount: parseFloat(this.state.amount, 10) * 100,
            createdAt: this.state.createdAt.valueOf(),
            note: this.state.note
        })
    }

    render(){
        return(
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input  
                        type="text" 
                        value={this.state.description} 
                        placeholder="description"
                        onChange={this.onDescriptionChange}
                        autoFocus
                    />
                    <input 
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onChangeAmount}
                    />
                    <SingleDatePicker
                        date = {this.state.createdAt}
                        onDateChange = {this.onDateChange}
                        focused = {this.state.focused}
                        onFocusChange = {this.onFocusChange}
                        numberOfMonths = {1}
                        isOutsideRange = {()=>false}
                    />
                    <textarea
                        placeholder="add a note for yout expense (optional)"
                        value = {this.state.note}
                        onChange = {this.onChangeNote}
                    ></textarea>
                    <button>{this.props.expense ? "Edit Expense":"Add Expense"}</button>
                </form>
            </div>
        )
    }
}


export default ExpenseForm;