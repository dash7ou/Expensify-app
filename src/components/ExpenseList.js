import React from "react";
import {connect} from "react-redux";
import ExpenseItem from "./ExpenseListItems";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = (props)=>(
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        {props.expenses.map(expense => <ExpenseItem key ={expense.id}  {...expense}/>)}
    </div>
)


const mapStateToProps = ({expenses, filters})=>{
    return {
        expenses: selectExpenses(expenses , filters),
        filters
    }
}


export default connect(mapStateToProps)(ExpenseList); 