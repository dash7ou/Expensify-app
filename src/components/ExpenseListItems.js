import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

const ExpenseItem = ({ description ,id, amount , createdAt})=>(

        <Link className="list-items" to={`/edit/${id}`}>
            <div>
                {description && <h3 className="list-items__title">{description}</h3>}
                <span className="list-item__sub-title"> {moment(createdAt).format('MMMM Do, YYYY')} </span>
            </div>
            <h3 className="list-item__data"> {numeral(amount /100).format('$0,0.00')} </h3>
        </Link>

)

export default ExpenseItem;