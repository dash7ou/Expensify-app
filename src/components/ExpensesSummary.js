import React from "react";
import {connect} from "react-redux";
import numeral from "numeral";
import selectorExpense from "../selectors/expenses";
import selectorExpensesToal from "../selectors/expensesTotal";


export const ExpensesSummary = ({expenseCount, expensesTotal})=>{
    const expensesWord = expenseCount === 1 ? "expense" : "expenses";
    const amountFormat = numeral(expensesTotal/ 100).format("$0,0.00");
    
    return (    
        <div>
            <h1>veiwing {expenseCount} {expensesWord} totalling {amountFormat}</h1>
        </div>
    )
}
const mapStateToProps =(state)=> {
    const visibleExpenses = selectorExpense(state.expenses, state.filters);

    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: selectorExpensesToal(visibleExpenses)
    }
}
export default connect(mapStateToProps)(ExpensesSummary);