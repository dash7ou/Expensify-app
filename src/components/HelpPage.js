import React from "react";

const HelpPage = ()=>(
    <div>
        this is our help page
    </div>
)

export default HelpPage;