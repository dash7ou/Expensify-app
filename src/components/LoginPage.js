import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from "../actions/auth";


export const LoginPage = ({startLogin})=>(
    <div className="login-page">
        <div className="login-page__box">
            <h1 className="login-page__title">Expensify App</h1>
            <p>It's time to get your expenses under control</p>
            <button onClick={startLogin}>Login</button>
        </div>
    </div>
)

const mapDispatchToProps = (dispatch)=>{
    return {
        startLogin: ()=> dispatch(startLogin())
    }
}

export default connect(undefined, mapDispatchToProps)(LoginPage);