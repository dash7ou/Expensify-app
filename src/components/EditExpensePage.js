import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import {editExpense , removeExpense} from "../actions/expenses"


export class EditExpensePage extends React.Component{
    onSubmit = (expense)=>{
        this.props.onSubmit(this.props.expense.id ,expense);
        this.props.history.push("/");
    }

    onRemove = ()=>{
        this.props.onRemove(this.props.expense);
        this.props.history.push("/");
    }

    render(){
        return(
            <div>
                <ExpenseForm
                    expense = {this.props.expense}
                    onSubmit = { this.onSubmit }
                />
                <button onClick = { this.onRemove }
                >
                    Remove
                </button>
            </div>
        )
    }
}


const mapStateToProps = (state , props)=>{
    return {
        expense: state.expenses.find(expense => expense.id === props.match.params.id)
    }
}

const mapDispatchToProps = (dispatch)=>{
    return {
        onSubmit: (id, expense)=> dispatch(editExpense(id, expense)),
        onRemove: (propsExpense) => dispatch(removeExpense(propsExpense))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);