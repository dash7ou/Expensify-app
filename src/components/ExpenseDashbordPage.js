import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpenseSummary from "./ExpensesSummary";

const ExpenseDashbordPage = ()=>(
    <div>
        <ExpenseSummary />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
)

export default ExpenseDashbordPage;