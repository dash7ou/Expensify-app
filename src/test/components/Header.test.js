import React from 'react';
import { shallow } from "enzyme";
import { Header } from "../../components/Header";

test("should render header correctly", ()=>{
    const wrapper = shallow(<Header startLogout={()=>{}} />);
    expect(wrapper).toMatchSnapshot();
})


test('should start logout when button click', ()=>{
    const startLogout = jest.fn();
    const wrapper = shallow(<Header startLogout={startLogout}/>);
    wrapper.find('button').simulate('click');
    expect(startLogout).toHaveBeenCalled();
})


 
//! old test 
// import ReactShallowRenderer from "react-test-renderer/shallow";
// test("should render header correctly", ()=>{
//     const renderer = new ReactShallowRenderer();
//     renderer.render(<Header />);
//     expect(renderer.getRenderOutput()).toMatchSnapshot();
// });