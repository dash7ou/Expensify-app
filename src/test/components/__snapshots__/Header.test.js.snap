// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render header correctly 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        activeClassName="is-active"
        className="header__title"
        replace={false}
        to="/dashbord"
      >
        <h1>
          Expensify
        </h1>
      </Link>
      <button
        className="button button--link"
        onClick={[Function]}
      >
        Logout
      </button>
    </div>
  </div>
</header>
`;
